### ISSUE TITLE
Script eval API enhancement

### ISSUE THREAD

**Issue by postcasio** [spheredev/neosphere#23]
I want to work on a new CommonJS-like module system - one of the requirements
for this is a way to evaluate scripts without access to the local scope. In
Node, this is done using `vm.runInThisContext`. Node additionally provides
some methods for creating new contexts and running scripts in them:
https://nodejs.org/api/vm.html

In Sphere, you can use EvaluateScript/RequireScript, but these functions do not
return the value, they can only work with paths, and they do not allow you to
change the context. Ideally, I would like a function that:
1. Takes a string to evaluate
2. Takes a filename to use in error messages
3. Takes an optional context
4. Returns the result

And a function that can create a context from an object.

e.g.

``` js
var script = 'myAbort("Test");';
var context = CreateContext({
    myAbort: Abort
});
EvaluateStringInContext(script, "test.js", context);
```

Taking a new context is not strictly necessary, but it is super useful. It can
be used to prevent the standard Sphere API being accessed by modules, which
I can then expose through a better designed module. It also means that
I can supply modules with the right `require`, `module`, `exports` etc values
through the context instead of hacking it by wrapping module scripts in
`(function(module, exports, __filename, __dirname) {})`.

**Comment by fatcerberus** [comment]
So essentially what you want is the ability to evaluate a script using a different object as the global object (just with some fancy jargon attached)...  I mean, I get the appeal, but it seems like a bit of overengineering, honestly.  And anyway, I feel it's outside the scope of minisphere, which is just designed to be a (mostly) compatible drop-in replacement for the aging official engine.  Wouldn't the Pegasus repo be a better place to post this?

That said, I'm pretty sure Duktape has a CommonJS system built in already:
http://duktape.org/guide.html#modules

Which means you could just use that if you really need the functionality.  I'm not likely to swap out Duktape anytime soon.

**Comment by postcasio** [comment]
> So essentially what you want is the ability to evaluate a script using a different object as the global object (just with some fancy jargon attached)... 

Yeah, that's pretty much it. I understand you don't want to add too much extra stuff but this does replace a feature from Sphere 1.5 that's currently missing - old SpiderMonkey supported taking a context as the second argument for `eval`.

I didn't realise Duktape had a CommonJS system built in however, I'll have a look at it later. If it works that will definitely be fine and save me a lot of trouble writing it myself!

**Comment by postcasio** [comment]
Looks like it works great. Using a simple:

``` coffee
Duktape.modSearch = (id) ->
  file = OpenRawFile('../scripts/src/' + id + '.js')
  throw new Error("Cannot find module #{id}") unless file

  script = file.read(file.getSize())
  file.close()

  CreateStringFromByteArray(script)
```

Thanks for your help!

**Comment by fatcerberus** [comment]
CoffeeScript, blegh. :-p. But that's awesome that that's all you need to implement modules, just have `modSearch` return a code string.

I'll agree that this is something worth considering codifying in the Sphere API, but that's a discussion for over on the Pegasus repo, methinks.

**Comment by fatcerberus** [comment]
So, I'm going to revive this thread to mention this:
https://github.com/fatcerberus/minisphere/tree/module-support

Built-in CommonJS module support, slated for inclusion in minisphere 1.2. :smile:

**Comment by fatcerberus** [comment]
CommonJS module support is implemented as of commit 2be48f55a9688f9977b163d9a30632f14ff890c9, so closing this issue now.
